kolibri/core/device/kolibri_plugin.py

```diff
@@ -64,3 +67,40 @@ def post_transfer(
                     "last_sync": time.time(),
                 },
             )
+
+
+class LearnerDeviceStatusOperation(KolibriVersionedSyncOperation):
+    version = "0.16.2"
+
+    def downgrade(self, context):
+        """
+        Delete LearnerDeviceStatus records that might cause an issue when syncing to a previous version.
+        For a single user sync, delete any learner device statuses associated with the single user.
+        For a facility sync, delete all learner device statuses for the facility.
+        :type context: morango.sync.context.LocalSessionContext
+        """
+        from kolibri.core.device.models import LearnerDeviceStatus
+
+        # get the user_id for the single user sync
+        # if it's not a single user sync, this will be None
+        user_id = get_user_id_for_single_user_sync(context)
+
+        # get the instance_id of the remote instance
+        instance_id = (
+            context.sync_session.client_instance_id
+            if context.is_server
+            else context.sync_session.server_instance_id
+        )
```

Comment:
So this `downgrade` method is called when this device is the producer of sync data. When it's a pull, then it must be the server if it's the producer, so restricting the learner status to only the `client_instance_id` device makes sense.

Although, when it's a push, in order for the device to only send one status, it can first consider whether the current device has created a sync status. If the current device has, we would still want to exclude the `client_instance_id`, because being a push producer (aka `not context.is_server`), the client's status is still what matters. If the current device hasn't created one, then we would want to delete all but the most recent user status.

So in summary, I don't think this code will function as expected when the device is pushing. For pulling, this looks okay.

When it's pushing without a user sub-partition, we would want to ensure the statuses are unique for each user, preferably using the most recent status for each user, meaning no explicit filtering on the instance. So the same approach for pulling doesn't quite work for pushing.

Reply:
Thanks @bjester - I didn't think I'd get it right first time. If you have any ideas on how to write a robust test case for this, so I can be sure of covering the bases, I'm very open to inspiration!

Reply:
Honestly coming back to the conversation about this, it was difficult to wrap my head around it! I realized the logic isn't straightforward at all :melting_face: 

I was wondering about testing. Since this relies on different versions, it doesn't seem straightforward either. The `KolibriVersionedSyncOperationTestCase` demonstrates how you could possibly set up a unit test. For sync integration tests, I believe it could be possible to set one up overriding `CUSTOM_INSTANCE_INFO` which influences the version that morango uses and communicates during syncing. https://github.com/learningequality/kolibri/blob/a2f580d6d065602f33ae1dccc1663d66e8f9e99c/kolibri/core/auth/test/test_morango_integration.py#L293